source/common/network/socket_option_impl.h

```diff
@@ -15,69 +16,115 @@ namespace Network {
+#define ENVOY_SOCKET_TCP_KEEPIDLE                                                                  \
+  Network::SocketOptionName(std::make_pair(IPPROTO_TCP, TCP_KEEPALIVE))
+#else
+#define ENVOY_SOCKET_TCP_KEEPIDLE Network::SocketOptionName()
+#endif
+
+#ifdef TCP_KEEPINTVL
+#define ENVOY_SOCKET_TCP_KEEPINTVL                                                                 \
+  Network::SocketOptionName(std::make_pair(IPPROTO_TCP, TCP_KEEPINTVL))
+#else
+#define ENVOY_SOCKET_TCP_KEEPINTVL Network::SocketOptionName()
+#endif
+
 #ifdef TCP_FASTOPEN
-#define ENVOY_SOCKET_TCP_FASTOPEN Network::SocketOptionName(TCP_FASTOPEN)
+#define ENVOY_SOCKET_TCP_FASTOPEN                                                                  \
+  Network::SocketOptionName(std::make_pair(IPPROTO_TCP, TCP_FASTOPEN))
 #else
 #define ENVOY_SOCKET_TCP_FASTOPEN Network::SocketOptionName()
 #endif
 
-class SocketOptionImpl : public Socket::Option, protected Logger::Loggable<Logger::Id::connection> {
+class SocketOptionImpl : public Socket::Option, Logger::Loggable<Logger::Id::connection> {
 public:
-  SocketOptionImpl(absl::optional<bool> transparent, absl::optional<bool> freebind)
-      : transparent_(transparent), freebind_(freebind) {}
+  SocketOptionImpl(Socket::SocketState in_state, Network::SocketOptionName optname, int value)
+      : in_state_(in_state), optname_(optname), value_(value) {}
 
   // Socket::Option
   bool setOption(Socket& socket, Socket::SocketState state) const override;
+
   // The common socket options don't require a hash key.
   void hashKey(std::vector<uint8_t>&) const override {}
 
-  /**
-   * Set a socket option that applies at both IPv4 and IPv6 socket levels. When the underlying FD
-   * is IPv6, this function will attempt to set at IPv6 unless the platform only supports the
-   * option at the IPv4 level.
-   * @param socket.
-   * @param ipv4_optname SocketOptionName for IPv4 level. Set to empty if not supported on
-   * platform.
-   * @param ipv6_optname SocketOptionName for IPv6 level. Set to empty if not supported on
-   * platform.
-   * @param optval as per setsockopt(2).
-   * @param optlen as per setsockopt(2).
-   * @return int as per setsockopt(2). ENOTSUP is returned if the option is not supported on the
-   * platform for fd after the above option level fallback semantics are taken into account or the
-   *         socket is non-IP.
-   */
-  static int setIpSocketOption(Socket& socket, SocketOptionName ipv4_optname,
-                               SocketOptionName ipv6_optname, const void* optval, socklen_t optlen);
+  bool isSupported() const;
+
+  static int setSocketOption(Socket& socket, Network::SocketOptionName optname, int value);
+
+private:
+  const Socket::SocketState in_state_;
+  const Network::SocketOptionName optname_;
+  const int value_;
+};
+
+class SocketOptionsImpl : public Socket::Option, Logger::Loggable<Logger::Id::connection> {
```

Comment:
Why do we need this given that `Socket` already has a collection of options? Doesn't this then become a collection of options for a collection of options?

Reply:
The two ideas I like are the last one (always use a special collection object) or for the factory to always return a vector/list (Socket::Options). I lean towards always using Socket::Options because then there isn't an issue with weird nesting.

With either of these approaches, I think eventually it would be better for each socketoption to have it's own object. You don't have to make that change as part of this PR if you don't want, but having 1 option per object, and a consistent way to dealing with a collection/set of these objects, would make this a bit cleaner.

Reply:
Yeah, I also like:
* One socket option per object.
* Factory returns always `Socket::Options`

Reply:
OK, I'll go that path, and for this PR I'll add a bunch of shared utility functions for managing these collections to avoid duplicating code in listener and upstream (rather than adding a new collection type with methods for the needed operations.)

Reply:
I've just pushed the change to get rid of SocketOptionsImpl. It would definitely look even cleaner with a new collection type, rather than static utility methods I added on Socket, but I think maybe this is already a sufficient improvement? I'm still missing test coverage as @htuch pointed out, which I'll fix tomorrow.